partition_snapshot_row_cursor.hh

```diff
@@ -111,52 +111,94 @@ class partition_snapshot_row_cursor final {
+    // is the lower_bound() of the current position in table schema order.
+    // For _reversed cursors it can be either lower_bound() in table order
+    // or lower_bound() in cursor's order, so should not be relied upon.
+    // if current entry is in the latest version then _latest_it points to it,
+    // also in _reversed mode.
     std::optional<mutation_partition::rows_type::iterator> _latest_it;
+
+    // Continuity corresponding to ranges which are not represented in _heap because the cursor
+    // went pass all the entries in those versions.
+    bool _background_continuity = false;
+
     bool _continuous{};
     bool _dummy{};
     const bool _unique_owner;
-    position_in_partition _position;
+    const bool _reversed;
+    position_in_partition _position; // table domain
     partition_snapshot::change_mark _change_mark;
 
+    position_in_partition_view to_table_domain(position_in_partition_view pos) const {
+        if (_reversed) [[unlikely]] {
+            return pos.reversed();
+        }
+        return pos;
+    }
+
+    position_in_partition_view to_query_domain(position_in_partition_view pos) const {
+        if (_reversed) [[unlikely]] {
+            return pos.reversed();
+        }
+        return pos;
+    }
+
+    struct version_heap_less_compare {
+        rows_entry::tri_compare _cmp;
+        partition_snapshot_row_cursor& _cur;
+    public:
+        explicit version_heap_less_compare(partition_snapshot_row_cursor& cur)
+            : _cmp(cur._schema)
+            , _cur(cur)
+        { }
+
+        bool operator()(const position_in_version& a, const position_in_version& b) {
+            auto res = _cmp(_cur.to_query_domain(a.it->position()), _cur.to_query_domain(b.it->position()));
+            return res > 0 || (res == 0 && a.version_no > b.version_no);
+        }
+    };
+
     // Removes the next row from _heap and puts it into _current_row
     bool recreate_current_row() {
         _current_row.clear();
+        _continuous = _background_continuity;
+        _dummy = true;
         if (_heap.empty()) {
+            if (_reversed) {
+                _position = position_in_partition::before_all_clustered_rows();
+            } else {
+                _position = position_in_partition::after_all_clustered_rows();
+            }
             return false;
         }
-        position_in_version::less_compare heap_less(_schema);
-        position_in_partition::equal_compare eq(_schema);
-        _continuous = false;
-        _dummy = true;
+        version_heap_less_compare heap_less(*this);
+        position_in_partition::equal_compare eq(*_snp.schema());
         do {
             boost::range::pop_heap(_heap, heap_less);
             memory::on_alloc_point();
             rows_entry& e = *_heap.back().it;
             _dummy &= bool(e.dummy());
-            _continuous |= bool(e.continuous());
+            _continuous |= bool(_heap.back().continuous);
```

Comment:
What's the difference between `_heap.back().continuous` and `_heap.back().it->continuous()`?

Reply:
Both flags determine continuity for the range in the direction of lower key values, the difference is in which schema domain is used for ordering. `_heap.back().continuous` is using cursor schema domain, so in reverse mode, the direction is towards higher key values in table schema domain, whereas `it->continuous()` is using table schema domain (it's as stored in memory, so schema domain is fixed).